uicomponent-compose/feed/src/main/java/io/github/droidkaigi/feeder/feed/FavoriteAnimation.kt

```diff
@@ -0,0 +1,138 @@
+import androidx.compose.ui.graphics.nativeCanvas
+import androidx.compose.ui.graphics.toArgb
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.util.fastForEach
+import androidx.core.content.ContextCompat
+import androidx.core.graphics.drawable.DrawableCompat
+import kotlin.random.Random
+
+private const val HEART_COUNT = 10
+private val moveInterpolator = FastOutSlowInEasing
+private val alphaInterpolator = FastOutLinearInEasing
+private val scaleInterpolator = LinearOutSlowInEasing
+
+@Composable
+fun FavoriteAnimation(
+    visible: Boolean,
+    modifier: Modifier = Modifier,
+) {
+    val transitionState = remember { MutableTransitionState(visible) }
+    val targetChanged = transitionState.targetState != visible
+    transitionState.targetState = visible
+    val transition = updateTransition(transitionState)
+    val animatedFraction by transition.animateFloat(
+        transitionSpec = {
+            tween(
+                durationMillis = if (visible) 400 else 0,
+                easing = LinearEasing
+            )
+        }
+    ) { if (it) 1f else 0f }
+
+    // workaround: reduce drawing count and block ugly blink when other items are toggled.
+    if (visible.not() || animatedFraction >= 1f) {
+        return
+    }
+
+    val hearts = remember { List(HEART_COUNT) { Heart() } }
+    if (targetChanged) {
+        hearts.fastForEach { heart ->
+            heart.reset()
+        }
+    }
+    FavoriteAnimationInternal(hearts, animatedFraction, modifier)
+}
+
+@Composable
+private fun FavoriteAnimationInternal(
+    hearts: List<Heart>,
+    fraction: Float,
+    modifier: Modifier = Modifier,
+) {
+    val moveProgress = moveInterpolator.transform(fraction)
+    val alphaProgress = alphaInterpolator.transform(fraction)
+    val scaleProgress = scaleInterpolator.transform(fraction)
+    hearts.fastForEach { heart ->
+        heart.transX = lerp(0.5f, heart.targetTransX, moveProgress)
+        heart.transY = lerp(1f, heart.targetTransY, moveProgress)
+        heart.alpha = lerp(1f, 0f, alphaProgress)
+        heart.scale = lerp(1f, 0.6f, scaleProgress)
+    }
+
+    val drawable = rememberHeartDrawable()
+    val drawableHalfWidth = drawable.intrinsicWidth / 2f
+    val drawableHalfHeight = drawable.intrinsicHeight / 2f
+
+    Canvas(modifier = modifier.fillMaxSize()) {
+        val width = size.width
+        val height = size.height
+        hearts.fastForEach { heart ->
+            val transX = lerp(drawableHalfWidth, width - drawableHalfWidth, heart.transX)
+            val transY = lerp(drawableHalfHeight, height - drawableHalfHeight, heart.transY)
+            translate(transX, transY) {
+                scale(scaleX = heart.scale, scaleY = heart.scale, pivot = Offset(0.5f, 0.5f)) {
+                    drawIntoCanvas { canvas ->
+                        drawable.alpha = (255 * heart.alpha).toInt()
+                        drawable.draw(canvas.nativeCanvas)
```

Comment:
I didn't know that we can use `canvas.nativeCanvas`!!

Reply:
Yes, we can use!